jbmc/src/java_bytecode/assignments_from_json.cpp

```diff
@@ -0,0 +1,843 @@
+};
+
+static java_class_typet
+followed_class_type(const exprt &expr, const symbol_table_baset &symbol_table)
+{
+  const pointer_typet &pointer_type = to_pointer_type(expr.type());
+  const java_class_typet &java_class_type =
+    to_java_class_type(namespacet{symbol_table}.follow(pointer_type.subtype()));
+  return java_class_type;
+}
+
+static bool
+has_enum_type(const exprt &expr, const symbol_table_baset &symbol_table)
+{
+  return followed_class_type(expr, symbol_table).get_is_enumeration();
+}
+
+/// This function is used as a workaround until reference-equal objects defined
+/// across several classes are tracked correctly. Once reference-equality works
+/// in all cases, this function can be removed.
+/// Until then, in the case of an enum expression that needs to be assigned a
+/// value, we distinguish between two cases:
+/// 1) the declaring class of the enum expression is the same as the type of the
+///    enum expression. For example, for an enum Pet {DOG, CAT}, the declaring
+///    class of the expression Pet.DOG is Pet, and the type of the expression is
+///    also Pet. The same is true for the expressions that represent the
+///    elements of the $VALUES array of Pet, and for any user-defined Pet-typed
+///    fields in Pet.java. In this case, initialize the expression just as a
+///    regular object that has known reference-equal objects. (Corresponds to
+///    creating the enum constant in Java or referencing it directly.)
+///    See assign_reference_from_json.
+/// 2) otherwise, initialize it by indexing the $VALUES array with the given
+///    ordinal. An example of this case would be the field `pet` in
+///    `class MyClass { Pet pet; }` (its declaring class is `MyClass` and its
+///    own type is `Pet`).
+///    See assign_enum_from_json.
+/// \param expr: an expression representing a Java object.
+/// \param symbol_table: used for looking up the type of \p expr.
+/// \param declaring_class_type: type of the class where \p expr is declared.
+/// \return `true` if \p expr has an enum type and is declared within the
+///   definition of that same type, `false` otherwise.
+bool is_enum_with_type_equal_to_declaring_type(
+  const exprt &expr,
+  const symbol_table_baset &symbol_table,
+  const java_class_typet &declaring_class_type)
+{
+  PRECONDITION(can_cast_type<pointer_typet>(expr.type()));
+  return followed_class_type(expr, symbol_table) == declaring_class_type &&
+         declaring_class_type.get_is_enumeration();
+}
+
+/// If the argument has a "@type" key, return the corresponding value, else
+/// return an empty optional.
+/// A runtime type that is different from the objects compile-time type should
+/// be specified in `json` in this way.
+/// Type values are of the format "my.package.name.ClassName".
+static optionalt<std::string> get_type(const jsont &json)
+{
+  if(!json.is_object())
+    return {};
+  const auto &json_object = to_json_object(json);
+  if(json_object.find("@type") == json_object.end())
```

Comment:
Does this comply with https://google.github.io/styleguide/jsoncstyleguide.xml ?

Reply:
Do you mean the `@`? It's directly taken from the `json-io` library, and in the Google style guide it says "JSON maps can use any Unicode character in key names."

Reply:
Depends on whether it is considered a `key` in a map or a `property name`.

Reply:
> property name - the name (or key) portion of the property.

Looks like keys and property names are the same thing?

Reply:
Not quite. A key is itself a piece of data,  eg a parameter name that is mapped to its type, whereas a property name is a name for a piece of data, eg 'zipCode' in an address object.